internal/controllers/v1beta1/garbagecollection/garbagecollection.go

```diff
@@ -0,0 +1,134 @@
+)
+
+type GarbageCollectionController struct {
+	name              string
+	certmanagerClient certmanagerversionedclient.Interface
+	istioClient       istioversionedclient.Interface
+	dryRun            bool
+}
+
+func NewGarbageCollectionController(istioClient istioversionedclient.Interface, certClient certmanagerversionedclient.Interface, opts ...OptionsFunc) *GarbageCollectionController {
+	gc := &GarbageCollectionController{
+		name:              "istio-garbage-collection-controller",
+		certmanagerClient: certClient,
+		istioClient:       istioClient,
+	}
+
+	for _, opt := range opts {
+		opt(gc)
+	}
+
+	return gc
+}
+
+func (c *GarbageCollectionController) SetupWithManager(ctx context.Context, mgr manager.Manager) error {
+	ctrl, err := controller.New(c.name, mgr, controller.Options{
+		Reconciler:  c,
+		RateLimiter: workqueue.NewItemExponentialFailureRateLimiter(time.Second, 1000*time.Second), // maxDelay is limited by the Informer defaultRsync interval
+	})
+	if err != nil {
+		return err
+	}
+
+	certmanagerInformerFactory := certmanagerinformers.NewSharedInformerFactoryWithOptions(c.certmanagerClient, time.Second*30, certmanagerinformers.WithTweakListOptions(func(listOptions *metav1.ListOptions) {
+		listOptions.LabelSelector = v1beta1labels.ManagedLabelSelector()
+	}))
+
+	if err := ctrl.Watch(&source.Informer{Informer: certmanagerInformerFactory.Certmanager().V1().Certificates().Informer()},
+		handler.Funcs{
+			// only handle Update so that Deleting a certificate does not trigger another Reconcile
+			// Create will also trigger an Update
+			UpdateFunc: updateFunc,
+		}); err != nil {
+		return err
+	}
+
+	certmanagerInformerFactory.Start(ctx.Done())
+
+	return nil
+}
+
+func (c *GarbageCollectionController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := log.FromContext(ctx)
+	log.V(1).Info("Debug")
+	log.V(1).Info("Running Garbage Collection Reconcile", "request", request.String())
+
+	certIface := c.certmanagerClient.CertmanagerV1().Certificates(request.Namespace)
+	cert, err := certIface.Get(ctx, request.Name, metav1.GetOptions{})
+	if err != nil {
+		log.Error(err, "failed to Get Certificate")
+		return reconcile.Result{
+			Requeue: true,
+		}, err
+	}
+
+	label := cert.Labels[v1beta1labels.ManagedLabel]
+	gatewayName, gatewayNamespace := v1beta1labels.ParseManagedLabel(label)
+
+	deleteCert := false
+	deleteOptions := metav1.DeleteOptions{}
+	if c.dryRun {
+		deleteOptions.DryRun = []string{"All"}
+	}
+
+	gateway, err := c.istioClient.NetworkingV1beta1().Gateways(gatewayNamespace).Get(ctx, gatewayName, metav1.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		log.V(1).Info("Gateway not found, marking Certificate for deletion", "gateway-namespace", gatewayNamespace, "gateway", gatewayName)
+		deleteCert = true
+	} else {
+		foundMatch := false
+
+		for _, s := range gateway.Spec.Servers {
+			if s.Tls.CredentialName == request.Name {
+				foundMatch = true
+				break
+			}
+		}
```

Comment:
nit: This can be extracted as a function.

Reply:
added `isCertificateInGatewaySpec` with unit test